src/VisualStudio/CSharp/Impl/CSharpPackage.cs

```diff
@@ -64,22 +65,24 @@ namespace Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService
 
     [ProvideAutomationProperties("TextEditor", "CSharp", Guids.TextManagerPackageString, profileNodeLabelId: 101, profileNodeDescriptionId: 106, resourcePackageGuid: Guids.CSharpPackageIdString)]
     [ProvideAutomationProperties("TextEditor", "CSharp-Specific", packageGuid: Guids.CSharpPackageIdString, profileNodeLabelId: 104, profileNodeDescriptionId: 105)]
-    [ProvideService(typeof(CSharpLanguageService), ServiceName = "C# Language Service")]
-    [ProvideService(typeof(ICSharpTempPECompilerService), ServiceName = "C# TempPE Compiler Service")]
+    [ProvideService(typeof(CSharpLanguageService), ServiceName = "C# Language Service", IsAsyncQueryable = true)]
+    [ProvideService(typeof(ICSharpTempPECompilerService), ServiceName = "C# TempPE Compiler Service", IsAsyncQueryable = true)]
     internal class CSharpPackage : AbstractPackage<CSharpPackage, CSharpLanguageService>, IVsUserSettingsQuery
     {
         private ObjectBrowserLibraryManager _libraryManager;
         private uint _libraryManagerCookie;
 
-        protected override void Initialize()
+        protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
         {
             try
             {
-                base.Initialize();
+                await base.InitializeAsync(cancellationToken, progress).ConfigureAwait(true);
+
+                await JoinableTaskFactory.SwitchToMainThreadAsync();
 
                 this.RegisterService<ICSharpTempPECompilerService>(() => new TempPECompilerService(this.Workspace.Services.GetService<IMetadataService>()));
 
-                RegisterObjectBrowserLibraryManager();
+                await RegisterObjectBrowserLibraryManagerAsync().ConfigureAwait(true);
```

Comment:
>ConfigureAwait(true); [](start = 65, length = 21)

Can't this last one be `ConfigureAwait(false)`? The only remaining code which can execute is the `FatalError.Report` and that can execute on any thread.

Reply:
> Can't this last one be `ConfigureAwait(false)`?

JTF aware code is encouraged to omit `ConfigureAwait` as a general development strategy. We have an analyzer that forbids this, so I explicitly coded the equivalent.

Reply:
It can be false even if it's explicit. 

---
In reply to: [187512611](https://github.com/dotnet/roslyn/pull/26787#discussion_r187512611) [](ancestors = 187512611)

Reply:
Given this is package Initialize, it's very much assumed everything is on the UI thread.

Reply:
Yes. I agree every line of code which has an effect should be on the UI thread. The `ConfigureAwait(true)` though is forcing the method to complete, while executing no other code, on the UI thread. That seems wasteful. Given that there is no more code to execute in the method, why not let it resume in the thread pool, possibly process an exception, and then complete.

Reply:
> The `ConfigureAwait(true)` though is forcing the method to complete, while executing no other code, on the UI thread. That seems wasteful.

There is a strong likelihood that the next initialization operation will switch to the main thread. The form seen here results in a slight reduction in latency for this case with almost no downside.

Reply:
The next initialization then would still need to call `SwitchToMainThreadAsync` for correctness. I agree if we've already done the work to put them on the UI thread then this would be cheaper. At the same time though I'm still confused how this makes latency lower. If there are no more init calls then we've switched to the UI thread for no reason. Switching back to the UI thread also means that we could allow an incorrectly constructed package work by accident.

I don't feel really strongly about this. But my intuition is this is wasteful (albeit minimally wasteful). If it's not minimally wasteful I'd like to understand why. Particularly because this scenario was given enough priority that we're going to be breaking people.

Reply:
The efficiency question of whether `ConfigureAwait(bool)` with `true` or `false` is better depends on how likely it is that this code will appear within a blocking callstack, such that it isn't actually async. If it's far more likely to truly be async, then `.ConfigureAwait(false)` is probably preferable when no more UI thread requirements exist in that method. 

But if it's reasonably likely that the code will run within a synchronously blocking callstack, then `.ConfigureAwait(true)` is preferred. The reason for this is you already are blocking a thread waiting for the result, and preserving the caller's context allows the continuation to execute on that same thread that otherwise has nothing better to do. If you use `ConfigureAwait(false)`, that forces the continuation to execute on yet another thread. In particular when the threadpool is fully occupied that increases latency and actually hurts performance. 

Whether or not it increases *your* latency, adding to the threadpool may contribute to threadpool starvation that impacts other people. Not that we should avoid using the threadpool when its the right choice -- you should absolutely use the threadpool for CPU intensive work that is free-threaded and thread-safe. But when we see UI delays in VS, it's most frequently that people are doing too much chunky work on the UI thread or because threadpool exhaustion is going on. I've *never* seen customer impacting UI delays because someone switched back to the UI thread at the tail end of their method instead of using the threadpool.

Given that, IMO, `.ConfigureAwait(true)` is appropriate for this, and most other JTF-aware code. I only explicitly switch off the UI thread when I'm about to do CPU intensive work that has no need for the UI thread. And in those cases, I find `await TaskScheduler.Default;` to be the most effective way to ensure the switch rather than use `.ConfigureAwait(false)` and *hope* that that switches me off the UI thread.